node/core/dispute-coordinator/src/tests.rs

```diff
@@ -3035,3 +3041,282 @@ fn participation_for_included_candidates() {
+			assert_eq!(votes.valid.raw().len(), 3); // 3 => 1 initial vote, 1 backing vote, and our vote
+			assert_eq!(votes.invalid.len(), 1);
+
+			// Wrap up
+			virtual_overseer.send(FromOrchestra::Signal(OverseerSignal::Conclude)).await;
+
+			test_state
+		})
+	});
+}
+
+/// Shows that when a candidate_included event is scraped from the chain we
+/// reprioritize any participation requests pertaining to that candidate.
+/// This involves moving the request for this candidate from the best effort
+/// queue to the priority queue.
+#[test]
+fn participation_requests_reprioritized_for_newly_included() {
+	test_harness(|mut test_state, mut virtual_overseer| {
+		Box::pin(async move {
+			let session = 1;
+			test_state.handle_resume_sync(&mut virtual_overseer, session).await;
+			let mut hash_info: HashMap<CandidateHash, (u8, CandidateReceipt)> = HashMap::new();
+
+			for repetition in 1..=20u8 {
+				// Building candidate receipts
+				let mut candidate_receipt = make_valid_candidate_receipt();
+				candidate_receipt.descriptor.pov_hash = Hash::from(
+					[repetition; 32], // Altering this receipt so its hash will be changed
+				);
+				let candidate_hash = candidate_receipt.hash();
+				hash_info.insert(candidate_hash, (repetition, candidate_receipt.clone()));
+
+				// Mark all candidates as backed, so their participation requests make it to best effort
+				test_state
+					.activate_leaf_at_session(
+						&mut virtual_overseer,
+						session,
+						1,
+						vec![make_candidate_backed_event(candidate_receipt.clone())],
+					)
+					.await;
+			}
+
+			for repetition in 1..=20u8 {
+				// Building candidate receipts
+				let mut candidate_receipt = make_valid_candidate_receipt();
+				candidate_receipt.descriptor.pov_hash = Hash::from(
+					[repetition; 32], // Altering this receipt so its hash will be changed
+				);
+				let candidate_hash = candidate_receipt.hash();
+				
+				// Create votes for candidates
+				let (valid_vote, invalid_vote) = generate_opposing_votes_pair(
+					&test_state,
+					ValidatorIndex(1),
+					ValidatorIndex(2),
+					candidate_hash,
+					session,
+					VoteType::Explicit,
+				)
+				.await;
+
+				// Import votes for candidates
+				virtual_overseer
+					.send(FromOrchestra::Communication {
+						msg: DisputeCoordinatorMessage::ImportStatements {
+							candidate_receipt: candidate_receipt.clone(),
+							session,
+							statements: vec![
+								(valid_vote, ValidatorIndex(1)),
+								(invalid_vote, ValidatorIndex(2)),
+							],
+							pending_confirmation: None,
+						},
+					})
+					.await;
+
+				// Meant to help process approval vote messages, though there's a chance early dispute
+				// distribution messages will come back first. We wait a very short time as to make sure
+				// we can drain the queue of incoming messages without accidentally draining the best
+				// effort queue, which we want filled.
+				println!("Message handle inside loop");
```

Comment:
Just an idea which I haven't tested at all.

At this point can we create a new function combining `handle_approval_vote_request` and `participation_with_distribution`? 
It will accept a set of hashes (for example) and will generate the necessary `GetApprovalSignaturesForCandidate` and `DisputeDistribution` responses. At the same time it will track the order at which the participations occur and will return in as a result. 

In the test you can call it and examine the result. (Of course you'll have to generate the candidate included event before calling the new function).

What do you think?

Reply:
That works nicely for ordering the `handle_approval_vote_request` and `participation_with_distribution` calls. I think the hang up with this approach would be finding a way to queue all the participation requests via `DisputeCoordinatorMessage::ImportStatements` initially. 

Through trial and error I've found that if you try to queue a second participation request without beforehand handling overseer messages for the first, then the second import times out and fails. 

The work around is to start handling overseer messages for each participation request as they happen. But we can't require any particular participation to run to completion before starting the next, as that would never allow us to fill the best effort queue as we need for testing. 

I still don't know if my work around will succeed, since there isn't another test in disputes dealing with this kind of raciness I can build off of. Below is the logging before the current error I'm encountering. Interestingly, if I increase the timeout of this `virtual_overseer.recv() past 150 milliseconds then each candidate is fully processed before the next is started. Then there is no error at all!

Error: (I changed some of the logging to be more descriptive)
 Running unittests src/lib.rs (/Users/paritybradley/Development/work_code/polkadot/target/debug/deps/polkadot_node_core_dispute_coordinator-5a5c906250bd86a4)

running 1 test
Handling overseer messages after importing candidate: 1
Got approval voting messages for candidate: 1
Handling overseer messages after importing candidate: 2
Availability recovery for candidate: 1
Request validation code
Got approval voting messages for candidate: 2
Validation triggered for candidate: 1
thread 'tests::participation_requests_reprioritized_for_newly_included' panicked at 'Test subsystem no longer live', /Users/paritybradley/Development/work_code/polkadot/node/subsystem-test-helpers/src/lib.rs:270:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test tests::participation_requests_reprioritized_for_newly_included ... FAILED

Reply:
I’ve come to the conclusion that our current disputes testing framework doesn’t support processing candidate events while participation is queued for any candidate. It would take a substantial rework of our virtual_overseer and test helper functions to more closely resemble the actual overseer in order to get around this. 

The current virtual overseer implementation can only do one thing at a time. Helper functions such as `activate_leaf_at_session` and `handle_approval_vote_request` rely on the assumption that the very next message handled by the overseer will be of a specific kind and pertaining to a specific candidate. Such assumptions do not hold in any case where we expect the overseer to do two things at once. For instance, when we want the overseer to activate a leaf containing a candidate included event while it is handling participation for a candidate in the best effort queue.

I suggest that if we want to test bumping a participation request from the best effort queue to the priority queue, then that test should be bundled with a PR which reworks the virtual overseer and its helper functions.